feat: OIDC authentication support

var-i opened the pull request:
<!--
Thank you for contributing to Chaos Mesh!

If you're unsure where to start, please refer to the contributing doc:

https://github.com/chaos-mesh/chaos-mesh/blob/master/CONTRIBUTING.md

If you still have questions, please let us know via issues.

Please follow https://www.conventionalcommits.org/en/v1.0.0/ when you open a new PR:
-->

## What problem does this PR solve?

<!-- Uncomment this line if some issues to close -->
Close #4141 

## What's changed and how it works?

<!-- Uncomment this line if this PR is associated with a proposal -->
Adds frontend and backend support for Kubernetes OIDC authentication

## Related changes

- [ ] This change also requires further updates to the [website](https://github.com/chaos-mesh/website) (e.g. docs)
- [ ] This change also requires further updates to the `UI interface`

## Cherry-pick to release branches (optional)

> This PR should be cherry-picked to the following release branches:

- [ ] release-2.6
- [ ] release-2.5

## Checklist

### CHANGELOG

> Must include at least one of them.

- [ ] I have updated the `CHANGELOG.md`
- [ ] I have labeled this PR with "no-need-update-changelog"

### Tests

> Must include at least one of them.

- [ ] Unit test
- [ ] E2E test
- [x] Manual test

### Side effects

- [ ] **Breaking backward compatibility**

## DCO

If you find the DCO check fails, please run commands like below (Depends on the actual situations. For example, if the failed commit isn't the most recent) to fix it:

```shell
git commit --amend --signoff
git push --force
```

var-i committed:
Initial OIDC support

Signed-off-by: Ivar Cmrecak <ivar.cmrecak@lego.com>
Signed-off-by: Zhiqiang ZHOU <im@strrl.dev>

g1eny0ung reviewed (commented):
@var-i Thanks for contributions! As this PR is implemented based on the GCP authentication, there are some areas that could be further refined and I will continue to give comments.

g1eny0ung commented on pkg/dashboard/apiserver/auth/oidc/utils.go:
```diff
@@ -0,0 +1,29 @@
+	"golang.org/x/oauth2"
+)
+
+func setCookie(c *gin.Context, token *oauth2.Token) {
```
The `setCookie` has already been defined in `gcp` package, it's better to extract a public method instead of making a copy of it.

g1eny0ung commented on pkg/dashboard/apiserver/auth/oidc/middleware.go:
```diff
@@ -0,0 +1,59 @@
+	ctx := c.Request.Context()
+
+	s.logger.Info("handling oidc middleware")
+	if c.Request.Header.Get("X-Authorization-Method") != "gcp" {
```
This part of the code should be updated. We can use `oidc` as the `X-Authorization-Method`. At the same time, we also need to update the frontend to adapt it.

g1eny0ung replied:
Or we can use `oauth` to unify all OAuth-based authorization.

STRRL commented:
It seems there have been a long time since the last update, I think I could take over this PR and push it get merged.

I would take a look about OIDC detail in chaos dashboard, and updates for the comments from @g1eny0ung

STRRL commented on pkg/dashboard/apiserver/auth/oidc/service.go:
```diff
@@ -0,0 +1,120 @@
+	url := *s.rootUrl
+	url.Path = path.Join(s.rootUrl.Path, "./api/auth/oidc/callback")
+
+	var Endpoint = oauth2.Endpoint{
+		AuthURL:  s.issuerUrl + "/oauth2/authorize",
+		TokenURL: s.issuerUrl + "/oauth2/token",
+	}
```
we can not use hardcoded endpoint here, because different providers have diffrenret conversions.

We could follow the configuration provided by the IdP like https://accounts.google.com/.well-known/openid-configuration, and maybe using https://github.com/coreos/go-oidc is a better choice.

reference: https://github.com/int128/kubelogin/blob/3f7513754c3e1bc118374980a8c38da323657a55/pkg/oidc/client/factory.go#L55

STRRL commented:
I found setting up a local dev env with OIDC would be very tricky, so I also write a doc for it: https://github.com/chaos-mesh/website/pull/418

STRRL committed:
feat: complete oidc exchange with custom CA

Signed-off-by: STRRL <im@strrl.dev>
Signed-off-by: Zhiqiang ZHOU <im@strrl.dev>

STRRL commented:
basically it could works now, but I am considering how to make it clean

STRRL commented:
Hi @g1eny0ung , I noticed we're using a  `<SyncIcon/>` here for the OIDC Login button.

https://github.com/chaos-mesh/chaos-mesh/blob/23f80553b7bdb69da179002c1effe3178ebfd03e/ui/app/src/components/TopContainer/Auth.tsx#L84-L95

Could we replace it with OIDC icon like this: https://fontawesome.com/v5/icons/openid?f=brands&s=solid?

I noticed that icons from MUI does not introduce that icon, do you have any suggestion?

g1eny0ung commented:
> Hi @g1eny0ung , I noticed we're using a `<SyncIcon/>` here for the OIDC Login button.
> 
> https://github.com/chaos-mesh/chaos-mesh/blob/23f80553b7bdb69da179002c1effe3178ebfd03e/ui/app/src/components/TopContainer/Auth.tsx#L84-L95
> 
> Could we replace it with OIDC icon like this: [fontawesome.com/v5/icons/openid?f=brands&s=solid](https://fontawesome.com/v5/icons/openid?f=brands&s=solid)?
> 
> I noticed that icons from MUI does not introduce that icon, do you have any suggestion?

@STRRL Sure. We can replace it with any logo in SVG format. But we can use https://www.svgrepo.com/svg/127268/openid-logo rather than font awesome icons because they must be attributed. This one is distributed under a non-copyright license.

If you agree, I can help to replace it.

STRRL commented:
> > Hi @g1eny0ung , I noticed we're using a `<SyncIcon/>` here for the OIDC Login button.
> > https://github.com/chaos-mesh/chaos-mesh/blob/23f80553b7bdb69da179002c1effe3178ebfd03e/ui/app/src/components/TopContainer/Auth.tsx#L84-L95
> > 
> > Could we replace it with OIDC icon like this: [fontawesome.com/v5/icons/openid?f=brands&s=solid](https://fontawesome.com/v5/icons/openid?f=brands&s=solid)?
> > I noticed that icons from MUI does not introduce that icon, do you have any suggestion?
> 
> @STRRL Sure. We can replace it with any logo in SVG format. But we can use [svgrepo.com/svg/127268/openid-logo](https://www.svgrepo.com/svg/127268/openid-logo) rather than font awesome icons because they must be attributed. This one is distributed under a non-copyright license.
> 
> If you agree, I can help to replace it.

For sure, I'm ok with it.

g1eny0ung committed:
chore: update OIDC icon

Signed-off-by: Yue Yang <g1enyy0ung@gmail.com>
Signed-off-by: Zhiqiang ZHOU <im@strrl.dev>

g1eny0ung commented:
@STRRL Already done in [a1e8a55](https://github.com/chaos-mesh/chaos-mesh/pull/4427/commits/a1e8a55fae6dc712932cd52c74088feed58c052d). You can take a look at the effect when you have time.

Andrewmatilde commented on pkg/dashboard/apiserver/auth/oidc/service.go:
```diff
@@ -0,0 +1,167 @@
+	caFilePath := conf.OidcCaFile
+	if caFilePath != "" {
+		caFile, err := os.Open(caFilePath)
+		if err != nil {
+			return nil, err
+		}
+		caFileBytes, err := io.ReadAll(caFile)
+		if err != nil {
+			return nil, err
+		}
+		caPEMBytes = caFileBytes
+		caFile.Close()
```
```suggestion
		defer caFile.Close()
		if err != nil {
			return nil, err
		}
		caFileBytes, err := io.ReadAll(caFile)
		if err != nil {
			return nil, err
		}
		caPEMBytes = caFileBytes
```

Andrewmatilde commented on pkg/dashboard/apiserver/auth/oidc/service.go:
```diff
@@ -0,0 +1,167 @@
+
+	oauth := s.getOauthConfig(c)
+
+	if s.caPEMBytes != nil {
```
I guess we need return NOT FOUND here.

STRRL committed:
Update pkg/dashboard/apiserver/auth/oidc/service.go

Co-authored-by: Andrewmatilde <davis6813585853062@outlook.com>
Signed-off-by: Zhiqiang ZHOU <im@strrl.dev>

STRRL committed:
chore: make check

Signed-off-by: Zhiqiang ZHOU <im@strrl.dev>

kushagra-paypay commented:
We also need this feature.

rafernandez commented:
It would be great if this feature were implemented.

KhaledELG commented:
up +1

STRRL commented:
It's better to have an integration test / e2e test before we merege it.

I would like to work in it.

STRRL committed:
fix(dashboard): handle OIDC provider init failure and fix auth method check

getOauthConfig dereferenced a nil provider when gooidc.NewProvider failed,
crashing the dashboard whenever the OIDC issuer was temporarily unreachable.
Return the error instead and let the redirect/callback/middleware handlers
surface it as an API error.

Also fix the OIDC refresh middleware comparing X-Authorization-Method against
"gcp" instead of "oidc", which meant the OIDC token refresh path never ran.

Signed-off-by: Zhiqiang ZHOU <im@strrl.dev>

STRRL committed:
test(oidc): add end-to-end integration test for dashboard OIDC auth

Provision a real Keycloak as the OIDC provider, patch the minikube
kube-apiserver to trust it, install Chaos Mesh with oidcSecurityMode enabled,
and drive the full authorization-code login flow through the dashboard. The
test asserts that Kubernetes RBAC honors the resulting id_token: rejected
without a binding, accepted after binding cluster-admin to the OIDC identity.

Runs in a dedicated integration_test_oidc.yml workflow, leaving the existing
integration_test.yml untouched.

Signed-off-by: Zhiqiang ZHOU <im@strrl.dev>

STRRL committed:
fix(dashboard): close OIDC CA file only after checking open error

The deferred caFile.Close() ran before the os.Open error check, so a failure
to open the configured OIDC CA file would dereference a nil *os.File and panic.
Move the error check ahead of the defer.

Signed-off-by: Zhiqiang ZHOU <im@strrl.dev>

STRRL committed:
chore(oidc): satisfy CI checks

- add Apache license headers to new yaml/workflow files
- add CHANGELOG entry for OIDC dashboard support
- regenerate values.schema.json (drop stale OIDCSecurityMode key)
- bump go-jose/v4 to v4.1.4 to fix GO-2026-4945 (CVSS 7.5)

Signed-off-by: Zhiqiang ZHOU <im@strrl.dev>

STRRL commented:
@codex @claude review please

STRRL committed:
fix(dashboard): keep OIDC refresh middleware auth method as "gcp"

The frontend reuses the GCP cookie path for OIDC sessions and still sends
X-Authorization-Method: "gcp". Changing only the backend middleware to "oidc"
broke front/back consistency and silently disabled OIDC token refresh, so
revert it to "gcp" to match the frontend.

Properly unifying this (or replacing the custom X-Authorization-* refresh
mechanism with an HttpOnly cookie read on the backend, as Kubernetes Dashboard
and Argo CD do) requires a coordinated frontend change and is deferred to a
follow-up auth refactor. The nil-provider panic and CA-file close-order fixes
are kept.

Signed-off-by: Zhiqiang ZHOU <im@strrl.dev>

STRRL committed:
fix(dashboard): address OIDC review findings

- service.go: guard the id_token type assertion with comma-ok so a provider
  that omits id_token returns a controlled API error instead of panicking;
  drop the Google-specific oauth2.ApprovalForce parameter
- middleware.go: remove the log line that printed full request headers
  (leaked the Authorization bearer token); demote the per-request log to V(1)
- chaos-dashboard-deployment.yaml: merge the two envFrom blocks so gcp and oidc
  existingSecret can coexist (duplicate YAML key was silently dropping one)
- handler.go: update the stale comment about middleware registration order

Signed-off-by: Zhiqiang ZHOU <im@strrl.dev>

STRRL commented:
OK for the rest of stuff, I will do it in anther PR.

STRRL commented:
I am going to merge this
